src/rt/arc.rs

```diff
@@ -121,21 +121,30 @@ impl State {
         assert_eq!(0, self.ref_cnt, "Arc leaked");
     }
 
-    pub(super) fn last_dependent_accesses<'a>(
-        &'a self,
+    pub(super) fn for_each_last_dependent_access(
+        &self,
         action: Action,
-    ) -> Box<dyn Iterator<Item = &'a Access> + 'a> {
+        mut f: impl FnMut(&Access),
+    ) {
         match action {
             // RefIncs are not dependent w/ RefDec, only inspections
-            Action::RefInc => Box::new([].into_iter()),
-            Action::RefDec => Box::new(self.last_ref_dec.iter()),
+            Action::RefInc => {}
+            Action::RefDec => self.last_ref_dec.iter().for_each(f),
         }
     }
```

Comment:
As an alternative, what do you think about something like this:
```rust
   pub(super) fn last_dependent_accesses<'a>(
        &'a self,
        action: Action,
    ) -> impl Iterator<Item = &'a Access> + 'a> {
        let iter = match action {
            Action::RefInc => None,
            Action::RefDec => Some(self.last_ref_dec.iter()),
        };
        iter.into_iter().flat_map(|i| i)
    }
```
Because the iterator constructed by `Option::flat_map` has the same type regardless of whether it's empty or not, we can return an `impl Iterator`, so we still have the performance benefit of not allocating, but we can continue returning an iterator. This lets downstream code consume it in more complex ways than `for_each`ing, which could be useful in the future.

This is just a thought, feel free to take it or leave it